### ISSUE TITLE
Use omg_rpc for omg_watcher (or not!)

### ISSUE THREAD

**Issue by InoMurko** [omisego/elixir-omg#559]
My assumption is that there are API routes that are specific to the Watcher. Which is ok and these routes need to be implemented in subsequent controllers.

I presume that because of DRY, things from :omg_rpc were pushed to into :omg_watcher, making things unclean (like `OMG.RPC.Web.Error`. If the functions here really need to be dried out, better to put them into a utility application. 

I do not see a use case why omg_rpc would need to be pulled into omg_watcher. I would rather suggest a refactor, where omg_rpc is named after it's main parent application, which is omg_api. And then completely clean up _mfas_ that are being pulled in. If omg_api is basically what childchain is, maybe naming it this way also makes sense! omg_api is basically not an API interface without omg_rpc.

**Comment by pnowosie** [comment]
I'll refer to the idea of `omg_rpc` which comes from the dependencies we had at the time. Now things looks a whole different so please keep in mind that the assumption of omg_rpc might get outdated or are wrong.

Because we had ch-ch & watcher so close to each other we started to think of them as a single app, so you could have a flag (or _mode_) which tells _mix_ you are willing to start ch-ch or watcher.
Then omg_rpc was planned to became web api layer serving both apps' endpoints. (You can have multiple Phx Endpoints in Phoenix app and Endpoint plugs Router, so we'd have ch-ch Endpoint pluging ch-ch related routes and watcher's one doing the same for watcher.) 

So as a natural consequence every web api related code that could work for both was declared in omg_rpc.   Taking a _utility library_ approach these things should be placed in the utilities.

**Comment by InoMurko** [comment]
Blocked by https://github.com/omisego/elixir-omg/issues/557

**Comment by InoMurko** [comment]
Looking at the exposed routes
```
elixir-omg  master $ mix phx.routes OMG.Watcher.Web.Router
       account_path  POST  /account.get_balance                       OMG.Watcher.Web.Controller.Account :get_balance
   transaction_path  POST  /account.get_transactions                  OMG.Watcher.Web.Controller.Transaction :get_transactions
       account_path  POST  /account.get_utxos                         OMG.Watcher.Web.Controller.Account :get_utxos
in_flight_exit_path  POST  /in_flight_exit.get_data                   OMG.Watcher.Web.Controller.InFlightExit :get_in_flight_exit
in_flight_exit_path  POST  /in_flight_exit.get_competitor             OMG.Watcher.Web.Controller.InFlightExit :get_competitor
in_flight_exit_path  POST  /in_flight_exit.prove_canonical            OMG.Watcher.Web.Controller.InFlightExit :prove_canonical
in_flight_exit_path  POST  /in_flight_exit.get_input_challenge_data   OMG.Watcher.Web.Controller.InFlightExit :get_input_challenge_data
in_flight_exit_path  POST  /in_flight_exit.get_output_challenge_data  OMG.Watcher.Web.Controller.InFlightExit :get_output_challenge_data
   transaction_path  POST  /transaction.all                           OMG.Watcher.Web.Controller.Transaction :get_transactions
   transaction_path  POST  /transaction.get                           OMG.Watcher.Web.Controller.Transaction :get_transaction
   transaction_path  POST  /transaction.submit                        OMG.Watcher.Web.Controller.Transaction :submit
   transaction_path  POST  /transaction.create                        OMG.Watcher.Web.Controller.Transaction :create
          utxo_path  POST  /utxo.get_exit_data                        OMG.Watcher.Web.Controller.Utxo :get_utxo_exit
     challenge_path  POST  /utxo.get_challenge_data                   OMG.Watcher.Web.Controller.Challenge :get_utxo_challenge
        status_path  POST  /status.get                                OMG.Watcher.Web.Controller.Status :get_status
      fallback_path  *     /*path                                     OMG.Watcher.Web.Controller.Fallback Route.NotFound

elixir-omg  master $ mix phx.routes OMG.ChildChain.Web.Router
** (Mix) the provided router, OMG.ChildChain.Web.Router, does not exist
elixir-omg  master $ mix phx.routes OMG.RPC.Web.Router
      block_path  POST  /block.get           OMG.RPC.Web.Controller.Block :get_block
transaction_path  POST  /transaction.submit  OMG.RPC.Web.Controller.Transaction :submit
   fallback_path  *     /*path               OMG.RPC.Web.Controller.Fallback Route.NotFound
```
what are the arguments to have OMG Watcher use OMG RPC? @pnowosie

**Comment by pnowosie** [comment]
Having in mind we havent't anticipated common `utils` module and web-api layer is just interface (at least Watcher can be used without http, see: `API` modules) it is just DRY to have all commob http related things in one place (e.g. input validation, response sanitization, serialization ...).

A lot has change from initial vision, but I can describe it and then we can list pros & cons.

__Vision__
Let's call this common application which serves http endpoints `HttpRPC` to avoid confussion. 
* `HttpRPC` depends on web frameworks (Phoenix, Cowboy, Plugs ...)
* HttpRPC provides swagger documentation to the web endpoints (it is outside)
* `HttpRPC` contains routes for `Child chain`, `Watcher security critical` & `Watcher informational` endpoints. 
* It implements controllers for these endpoints. Let's not focus on Phoenix details here and just say that each controller serves http related functionality (retrieves parameters, validates them, transforms & serialize reponse). 
* Real work (not http related) is delegated to `API` modules (`backing controller pattern`), APIs implements the same named functions as controllers but receive parameters from elixir world, and fails quickly if parameters are invalid. (See how Watcher controllers implement it for example).

API modules are implemented outside `HttpRPC` module, regarding theirs functionality in `Child chain` or `Watcher` apps.

`Child chain` & `Watcher` application depends on `HttpRPC` (reversed dependency - so Http layer is just one of the children of the app). Let say `Child chain` starting `HttpRPC` application provides some flags to ignore routes belonging to `Watcher` and vice-versa. This also allows to build functionality on top of Watcher and don't start `HttpRPC` and call `Watcher` via `API` layer.

Because we reversed dependency we could move `HttpRPC` outside of the umbrella and treat it as separate interface layer of `Child chain` & `Watcher`. 

What do you think about it @pdobacz @InoMurko - am I missed something? I do appreciate Elixir experts view of the topic! Thanks

**Comment by InoMurko** [comment]
I think I understand what you mean @pnowosie. My fear is that we would gain very little by doing this.
What do Child chain and Watcher share in terms of implementation? My understanding is they share everything that's jsonrpc implementation related (request and response validation, encoding and decoding). Is there anything else I might be missing?

Having both in the same boat also increases the change of breaking things for one and the other - though we're not really leveraging much of Phoenix that we couldn't do by having a very simple `Plug.Cowboy` implementation (like Mana does for https://github.com/mana-ethereum/mana/tree/master/apps/jsonrpc2/lib/jsonrpc2 for protocols http, websockets and IPC). So DRY is good, but not where it decreases separation.

**Comment by pnowosie** [comment]
Agreed regarding new modules boundaries (API endpoint for Watcher in Watcher app (as is), API endpoint merged into omg_child_chain and common web-utilities in separate module used by both APIs).

(Maybe I got you wrong) If you're proposing to ditch Phoenix dependency and re-implement web-layer with self-made module - I believe it isn't worth the effort.

Thanks!

**Comment by InoMurko** [comment]
Got another idea we could do:
we could use https://hexdocs.pm/phoenix/Phoenix.Router.html#forward/4 `Router.forward/4`.
Basically, the idea is:
We have one Phoenix application called `omg_rpc` and three Routers + and request and response validation, encoding and decoding..., but *NO* logic. All business logic is in the application that has the router!
1. Top level Root router. Let's call it `RootRouter`
2. Watcher router. `WRouter` in `omg_watcher`
3. Child Chain router. `ChRouter` in `omg_child_chain`

Watcher flow would go like this (Ch Ch would be similar): 
- Application `omg_watcher` boots and configures omg_rpc via `Application.put_env` and sets the `forward_to_router` option on `RootRouter omg_rpc` to `WRouter`.
- `omg_rpc` receives a request, the request goes to `RootRouter` that checks witch router was set as the _handler_, and forwards the request to `WRouter`. `WRouter` in `omg_watcher` has all the route handlers implemented, and returns the response.

**Comment by InoMurko** [comment]
This is a similar approach https://blog.appsignal.com/2019/04/16/elixir-alchemy-routing-phoenix-umbrella-apps.html

**Comment by pnowosie** [comment]
Here are my 2¢

I think the latest proposition could make a project harder to configure and gives no gains (comparing to your prev proposition with 2 Phx apps and shared web-utils).
We still need to depend on Phoenix in Ch-ch & W and have Router + Controllers + Views defined in each app.
Also, currently our web-layer has no logic by itself and delegate work to API modules.

Also docs says that: `The forwarded plug will be initialized at compile time.` so the new `omg_rpc` would need to depend on Ch-ch & Watcher. (:question:)

Despite the critics I appreciate you think about alternatives.

**Comment by InoMurko** [comment]
Hey,
yeah, the appsignal link is not to be followed closely, because their phoenix app is meant to be used by every umbrella app. What we need is to be able to separate two use cases but try to make it reusable/pluggable. In this case, my proposal https://github.com/omisego/elixir-omg/issues/559#issuecomment-483820013 is closer to our needs.

**Comment by InoMurko** [comment]
Done.
